### PR TITLE
Add Stopping Criteria for loop

### LINKED ISSUE #243: Stopping criterion API
[Discussion](https://github.com/baal-org/baal/discussions/242#discussioncomment-4197328)
[Related Paper](https://arxiv.org/pdf/2104.01836.pdf)

I see this as a new object we include with the following strategies:
* If the average uncertainty is below a threshold
* If the labeling budget has been exhausted
* If the model has converged.

### PR DESCRIPTION
**Pull request by Dref360**
## Summary:

Allow the users to stop the loop according to various parameters.

### Features:
Fixes #243 

## Checklist:

* [ ] Your code is documented (To validate this, add your module to `tests/documentation_test.py`).
* [ ] Your code is tested with unit tests.
* [ ] You moved your Issue to the PR state.

### CODE REVIEW

File: experiments/mlp_mcdropout.py [hunk 1]
```diff
@@ -54,16 +55,18 @@
...
         al_dataset, optimizer=optimizer, batch_size=32, epoch=10, use_cuda=use_cuda
     )
     test_loss = wrapper.test_on_dataset(test_ds, batch_size=32, use_cuda=use_cuda)
 
     pprint(wrapper.get_metrics())
-    flag = al_loop.step()
-    if not flag:
-        # We are done labelling! stopping
+    al_loop.step()
+    if stopping_criterion.should_stop():
```
Comment (parmidaatg) [inline experiments/mlp_mcdropout.py:71]: Wondering if you wanna change these two lines to below lines  to account for general exhaustion of the dataset.  The scenario does not apply to this example but as a demonstration of how to account for general exhaustion of the dataset and if the stop criteria is a metric limit

```suggestion
    flag = al_loop.step()
    if stopping_criterion.should_stop() or flag:
```

**Review by parmidaatg — COMMENTED** [review]
neat, Just one suggestion but not necessary.

File: experiments/mlp_mcdropout.py (hunk 1 above)
Comment (Dref360) [reply experiments/mlp_mcdropout.py:71]: My thinking was that stopping criteria must be added to ALLoop which would check for both the criteria and exhaustion. It would require ALLoop to know about metrics which it can't do right now.

That would be a major breaking change, so we might want to have a new object instead and deprecate ALLoop :thinking: and then this new object would do the entire experiment?

**Comment by Dref360** [comment]
I'll merge this and then I'll make a proposal for the new object that replaces ActiveLearningLoop into an object that can uses this new Criterion to automatically stop an experiment.

**Review by parmidaatg — APPROVED** [review]
sounds good, lets go
